definitions/npm/styled-components_v2.x.x/flow_v0.53.x-/styled-components_v2.x.x.js

```diff
@@ -0,0 +1,369 @@
+// @flow
+
+type $npm$styledComponents$Interpolation = (<C: {}>(executionContext: C) => string) | string | number;
+type $npm$styledComponents$NameGenerator = (hash: number) => string;
+
+type $npm$styledComponents$TaggedTemplateLiteral<R> = {| (Array<string>, $npm$styledComponents$Interpolation): R |};
+
+// ---- FUNCTIONAL COMPONENT DEFINITIONS ----
+type $npm$styledComponents$ReactComponentFunctional<Props, DefaultProps = *> = {
+  (Props): React$Element<*>,
+  defaultProps: DefaultProps,
+}
+
+type $npm$styledComponents$ReactComponentFunctionalUndefinedDefaultProps<Props> = { (Props): React$Element<*> }
+
+// ---- CLASS COMPONENT DEFINITIONS ----
+class $npm$styledComponents$ReactComponent<Props, DefaultProps> extends React$Component<Props> {
+  static defaultProps: DefaultProps
+}
+type $npm$styledComponents$ReactComponentClass<Props, DefaultProps = *> = Class<$npm$styledComponents$ReactComponent<Props, DefaultProps>>
+type $npm$styledComponents$ReactComponentClassUndefinedDefaultProps<Props> = Class<React$Component<Props>>
+
+// ---- COMPONENT FUNCTIONS INPUT (UNION) & OUTPUT (INTERSECTION) ----
+type $npm$styledComponents$ReactComponentUnion<Props> =
```

Comment:
Same as #1156, is there a problem with `React$ComponentType`?

Reply:
IIRC, ``React$ComponentType`` did not exist when I wrote this type originally.

This definition references the use of defaultProps, whereas ``React$ComponentType`` does not.

```js
// ---- FUNCTIONAL COMPONENT DEFINITIONS ----
type $npm$styledComponents$ReactComponentFunctional<Props, DefaultProps = *> = {
  (Props): React$Element<*>,
  defaultProps: DefaultProps,
}
```
[#L8](https://github.com/mwalkerwells/flow-typed/blob/0b783539a7d1fda556b973c01c3289f7875be938/definitions/npm/styled-components_v2.x.x/flow_v0.53.x-/styled-components_v2.x.x.js#L8)

``$npm$styledComponents$ReactComponentUnion<Props>`` should be contravariant WRT ``React$ComponentType``, which should make it more flexible, but still safe.

That being said, we should update the definition:
```js
// ---- COMPONENT FUNCTIONS INPUT (UNION) & OUTPUT (INTERSECTION) ----
type $npm$styledComponents$ReactComponentUnion<Props> =
  | $npm$styledComponents$ReactComponentFunctional<Props>
  | $npm$styledComponents$ReactComponentFunctionalUndefinedDefaultProps<Props>
  | $npm$styledComponents$ReactComponentClass<Props>
  | $npm$styledComponents$ReactComponentClassUndefinedDefaultProps<Props>
  | React$StatelessFunctionalComponent<Props>; // ADDED
```
...and this:
```js
type $npm$styledComponents$ReactComponentClassUndefinedDefaultProps<Props> = Class<React$Component<Props, *>>
```
^^ without this, the  [default ``State`` is void](https://github.com/facebook/flow/blob/master/lib/react.js#L29), & we want it to be inferred.

Reply:
> IIRC, React$ComponentType did not exist when I wrote this type originally.

It's only been out since `flow@0.53`, I wouldn't be surprised.
> This definition references the use of defaultProps, whereas React$ComponentType does not.

What benefits does `defaultProps` provide here?

Reply:
Tried to summarize it here (https://github.com/flowtype/flow-typed/pull/1156#discussion_r136143193 / https://github.com/flowtype/flow-typed/pull/1156#discussion_r136144500), but I can probably do better...

If you think of a React component as a callable object that takes props and context as input and returns a ``React$Node``, and has a ``defaultProps`` property...a HOC (a function that takes and returns an object), is merely a callable object morphism, typically transforming ``Props`` and ``DefaultProps``.

Reply:
```js
declare var Component: {
  ({ foo: string }): React$Node,
  defaultProps: {},
}

declare var HOC: ({
  ({ foo: string }): React$Node,
  defaultProps: {},
}) => ({
  ({}): React$Node,
  defaultProps: { foo: string },
})

const Test = HOC(Component)

<Test /> // No ``foo`` prop
```

Reply:
That's why use ``DefaultProps`` in my definition, because it's a lossy type morphism without it.

Reply:
While [it isn't explicitly typed](https://github.com/facebook/flow/blob/v0.54.0/lib/react.js#L81-L86), the intention of `defaultProps` seems to be that it'd be a subtype of `Props`.

Reply:
Hmmm... you mean that ``DefaultProps`` should be a supertype of ``Props``? I had read [that](https://github.com/facebook/flow/blob/v0.54.0/lib/react.js#L81-L86) before, but after re-reading it, assuming you meant supertype,  I think you're right...